Proxy plugins

kugel- opened the pull request:
This is the patch set that implements proxy plugins as per mailing list thread a few months ago.

Please see commit a006f710 ("plugins: add geany_plugin_register_proxy() to the plugin API") which contains a detailed description via API documentation.

Short summary: Plugins can register as proxy plugins. They register a list of file extensions and a few callbacks. Then, when Geany scans for normal plugins it will also look for these new file types, and call the proxy plugin's callbacks instead of the built-in shared library loader. The proxy will ultimately call geany_plugin_register() on behalf of the sub-plugin.

To the user, these sub-plugins are indistinguishable except the Plugin Manager indicates the relationship to the proxy via a TreeView. Within Geany, they are treated as first-class citizens.

Please review (and ideally merge).

kugel- commented:
I don't know where to PR'ify but here's a converted GeanyPy acting as a real-word showcase: https://github.com/kugel-/geany-plugins/tree/geanypy-pluxy

kugel- commented:
@b4n @elextr @codebrainz sorry to annoy you but I really really want this for 1.26 so give it a look pretty please

codebrainz commented:
@kugel- this weekend is a long-weekend here, if you ping me on Saturday (I'm about half a day behind your timezone), I will try to sit down with this and at least give it some testing/reviewing.

It looks to have a build error for [this newish function](https://developer.gnome.org/glib/stable/glib-Pointer-Arrays.html#g-ptr-array-insert) on all platforms, according to Travis.

b4n commented:
> @b4n @elextr @codebrainz sorry to annoy you but I really really want this for 1.26 so give it a look pretty please 

I'm busy this WE, but I should be on it starting next Monday.  Feel free to ping me continuously if I don't ;)

codebrainz commented on plugins/demoproxy.c:
```diff
@@ -0,0 +1,218 @@
+ *      demoproxy.c - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2007-2012 Enrico Tröger <enrico(dot)troeger(at)uvena(dot)de>
+ *      Copyright 2007-2012 Nick Treleaven <nick(dot)treleaven(at)btinternet(dot)com>
```
Since you wrote this new file, it would be more appropriate to put your copyright in there.

elextr replied:
Agree this should be copyright @kugel-

elextr commented:
@kugel- all travis builds show an implicitly defined function.  Did you miss an include file?

b4n commented:
Here's what GCC tells me when I build this:
```gcc
../../src/plugins.c: In function 'proxied_count_inc':
../../src/plugins.c:140:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
 ^
../../src/plugins.c: In function 'proxied_count_dec':
../../src/plugins.c:149:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
 ^
../../src/plugins.c: In function 'plugin_load_so':
../../src/plugins.c:590:14: warning: unused variable 'p_geany_data' [-Wunused-variable]
  GeanyData **p_geany_data;
              ^
../../src/plugins.c: In function 'plugin_new':
../../src/plugins.c:648:11: warning: unused variable 'module' [-Wunused-variable]
  GModule *module;
           ^
../../src/plugins.c: In function 'unregister_proxy':
../../src/plugins.c:897:110: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
  foreach_ptr_array(p, i, supported_plugin_list)
                                                                                                              ^
../../src/plugins.c: In function 'is_plugin':
../../src/plugins.c:1022:114: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
  foreach_ptr_array(proxy, i, supported_plugin_list)
                                                                                                                  ^
../../src/plugins.c: In function 'load_plugins_from_path':
../../src/plugins.c:1086:17: warning: unused variable 'tmp' [-Wunused-variable]
  gchar *fname, *tmp;
                 ^
../../src/plugins.c: In function 'pm_plugin_toggled':
../../src/plugins.c:1432:29: warning: unused variable 'valid' [-Wunused-variable]
  gboolean old_state, state, valid;
                             ^
../../src/plugins.c: In function 'pm_treeview_query_tooltip':
../../src/plugins.c:1583:24: warning: initialization discards 'const' qualifier from pointer target type
   gchar *uncheck_str = "\n<i>Other plugins depend on this. Disable them first to allow deactivation.</i>\n";
                        ^
../../src/plugins.c:1590:3: warning: format not a string literal, argument types not checked [-Wformat-nonliteral]
   markup = g_strdup_printf(tmp, can_uncheck ? "" : uncheck_str);
   ^
../../src/plugins.c: In function 'pm_prepare_treeview':
../../src/plugins.c:1738:9: warning: unused variable 'list' [-Wunused-variable]
  GList *list;
         ^
../../src/plugins.c:1737:14: warning: unused variable 'iter' [-Wunused-variable]
  GtkTreeIter iter;
              ^
../../src/plugins.c: At top level:
../../src/plugins.c:78:22: warning: 'builtin_extensions' defined but not used [-Wunused-variable]
 static const gchar * builtin_extensions[] = { G_MODULE_SUFFIX, NULL };
                      ^
```
```
../../plugins/demoproxy.c: In function 'proxy_init':
../../plugins/demoproxy.c:68:3: warning: format not a string literal, argument types not checked [-Wformat-nonliteral]
   gchar *key = g_strdup_printf(fmt, i++);
   ^
../../plugins/demoproxy.c: In function 'proxy_help':
../../plugins/demoproxy.c:92:12: warning: unused variable 'file' [-Wunused-variable]
  GKeyFile *file;
            ^
```

b4n commented:
I find the `geany_plugin_register_proxy()` API a bit odd: the proxy has to call this, yet fill the vfuncs/callback in the `GeanyPlugin` structure.  This seem odd API-wise to me, and seem to lead to a seemingly arbitrary limitation of one proxy per native plugin.
Is there a particular reason why it's done that way?

kugel- commented:
The separate API allows to do the registration in the plugin's init(), and have the file extensions deppend on plugin configuration. Also a API-wise separattion to geany_plugin_register() is needed to allow for nested proxies.

A proxy can be a proxy for many types. the only limitation is that only one set of ProxyFuncs is supported per proxy. But this limitation is not a problem, since the proxies can do their own dispatching. For geany it simplifies storing the pointers (avoids introducing a list of vtables, etc)

b4n commented on src/plugins.c:
```diff
@@ -830,6 +1000,52 @@ static gboolean check_plugin_path(const gchar *fname)
+	if (tmp == NULL)
+		return FALSE;
+	/* ensure the dot is really part of the filename */
+	else if (strchr(tmp, G_DIR_SEPARATOR) != NULL)
```
On Windows we probably also need to check `/`

kugel- replied:
Oh look. I used strchr(). I'm inconsistent with myself :/

Regarding the /, do we have to do this everywhere? I didn't notice an #ifdef hell about that yet

b4n replied:
We don't deconstruct paths manually very often, and would generally use `g_path_get_basename()`.  But we do have [`utils_tidy_path()`](https://github.com/geany/geany/tree/master/src/utils.c#L1795) that does it.  I'm not certain we need it here, but I could easily imagine it could happen.

b4n commented on plugins/demoproxy.c:
```diff
@@ -0,0 +1,218 @@
+	gchar linebuf[128];
+	FILE *f = fopen(filename, "r");
+	fgets(linebuf, sizeof(linebuf), f);
+	match = utils_str_equal(linebuf, "#!!proxy_MAGIC!!\n");
```
the magic doesn't match the one in the docs or the one actually used in the sample *.px*

kugel- replied:
Sorry, superfluous search & replace in action.

b4n commented on plugins/demoproxytest.px:
```diff
@@ -0,0 +1,15 @@
+#!!PLUXY_MAGIC!!
```
doesn't match the one in the docs, nor the one actually checked by the demoproxy

kugel- commented:
@b4n @codebrainz @elextr I hopefully addressed all remarks, please review again (I'll take care of the travis issue later on)

kugel- commented:
I forced pushed to some fixup commits since those haven't been commented on yet.

b4n commented on plugins/demoproxy.c:
```diff
@@ -0,0 +1,207 @@
+ *
+ * Note: This is not installed by default, but (on *nix) you can install it as follows:
+ *
+ * Compile geany.Then copy or symlink the plugins/demoproxy.so and plugins/demoproxytest.px files
```
missing space between sentences

b4n commented on src/plugins.c:
```diff
@@ -168,22 +216,28 @@ static Plugin *find_active_plugin_by_name(const gchar *filename)
 plugin_check_version(Plugin *plugin, int plugin_version_code)
 {
-	GModule *module = plugin->module;
+	gchar *name = NULL;
```
unused (shadowed in inner scopes)

b4n replied:
still not used

b4n commented on src/plugins.c:
```diff
@@ -747,32 +876,86 @@ plugin_cleanup(Plugin *plugin)
+}
+
+
+/* Returns true when the removal was successful (=> never for non-pluxies) */
```
no biggie, but there's an occurrence of "pluxy" left here

b4n commented on src/plugins.c:
```diff
@@ -747,32 +876,86 @@ plugin_cleanup(Plugin *plugin)
+
+
+/* Returns true when the removal was successful (=> never for non-pluxies) */
+static gboolean
```
style stuff, if you feel like it

b4n commented on src/plugins.c:
```diff
@@ -830,25 +1013,82 @@ static gboolean check_plugin_path(const gchar *fname)
-
-		if (!EMPTY(fname) && g_file_test(fname, G_FILE_TEST_EXISTS))
+		proxies = active_proxies.length;
+		if (failed_plugins_list != NULL)
```
test is unnecessary as `NULL` is a valid GList

b4n commented on src/plugins.c:
```diff
@@ -858,18 +1098,19 @@ load_plugins_from_path(const gchar *path)
 	GSList *list, *item;
 	gchar *fname, *tmp;
```
`tmp` becomes unused

b4n replied:
while at it, moving the declaration of `fname` in the inner scope would be perfect :)

b4n commented on src/plugins.c:
```diff
@@ -858,18 +1098,19 @@ load_plugins_from_path(const gchar *path)
 	GSList *list, *item;
 	gchar *fname, *tmp;
 	gint count = 0;
+	PluginProxy *proxy;
```
(this one too)

b4n commented on src/plugins.c:
```diff
@@ -858,18 +1098,19 @@ load_plugins_from_path(const gchar *path)
+		proxy = is_plugin(fname);
+		if (proxy != NULL)
+		{
+			if (plugin_new(proxy->plugin, fname, FALSE, TRUE))
```
those two tests could be combined in a single `if()` (no biggie)

b4n commented on src/plugins.c:
```diff
@@ -1028,6 +1298,17 @@ void plugins_init(void)
+}
+
+
+/* Same as plugin_free(), except it does nothing for pluxies-in-use, to be called on
```
s/pluxies/proxies/

b4n commented on src/plugins.c:
```diff
@@ -1137,6 +1419,27 @@ static void pm_selection_changed(GtkTreeSelection *selection, gpointer user_data
+	gboolean valid;
+
+	for (valid = gtk_tree_model_get_iter_first(model, iter);
+	     valid && p;
```
which check `p` in the loop?  it can't change if I'm not mistaken

kugel- replied:
Tried to save LOCs by avoiding an extra if ()  :-)

kugel- replied:
Actually it can't be NULL at all

b4n commented on src/plugins.c:
```diff
@@ -1186,16 +1495,89 @@ static void pm_plugin_toggled(GtkCellRendererToggle *cell, gchar *pth, gpointer
+			GtkTreeIter parent;
+			gboolean can_uncheck;
+			GtkTreePath *store_path = gtk_tree_model_filter_convert_path_to_child_path(
+			                                GTK_TREE_MODEL_FILTER(model), path);
```
are we sure the path is still the same?  I could imagine not if e.g. the plugin changed name upon reload maybe

kugel- replied:
The rest of the function assumes the same. I think it's nearly impossible that it changed, since the tree was just created when opening the PM dialog (i.e. it's not cached from the startup).

b4n replied:
hum nevermind my remark, I assumed (incorrectly) the row was removed and re-added, which could change the layout of the data store.  But actually it's not, the existing one is only updated, so it's totally fine.
sorry for the noise

b4n commented on src/plugins.c:
```diff
@@ -1186,16 +1495,89 @@ static void pm_plugin_toggled(GtkCellRendererToggle *cell, gchar *pth, gpointer
+	gtk_tree_path_free(path);
 	g_free(file_name);
 }
 
```
style: missing newline :]

b4n commented on src/plugins.c:
```diff
@@ -1186,16 +1495,89 @@ static void pm_plugin_toggled(GtkCellRendererToggle *cell, gchar *pth, gpointer
+		Plugin *p;
+		for (; list != NULL; list = list->next)
+		{
+			p = list->data;
```
declaration in the loop

b4n replied:
(OK I see it was like this in the original code)

kugel- replied:
Like for (Plugin *p = list->data; list != NULL; list = list->next, p = list ? list->data : NULL)?

I'm not sure that's an improvement really.

b4n replied:
No I mean `Plugin *p = list->data;` in the loop body* sorry

b4n replied:
no biggie but still applies if you'd like to change it

b4n commented:
> The separate API allows to do the registration in the plugin's init(), and have the file extensions deppend on plugin configuration. Also a API-wise separattion to geany_plugin_register() is needed to allow for nested proxies.

I get that, my point was that I would find the API less odd if one filled a specific structure, rather than one already used for something else.

e.g.
```C
Proxy proxy = {
  .extensions = { "so", NULL },
  .load = my_load,
  .probe = my_probe,
  .unload = my_unload
};
geany_plugin_register_proxy(plugin, &proxy);
```
or alike.  But I know it has drawbacks, like binary incompatibility if the struct expands (solvable by  versioning the struct, but a bit annoying), etc.

I don't really mind though, as it does has a few advantages (at least being simple and easy to expand), and that API won't be used by everyone so any oddness won't affect too many people.

> A proxy can be a proxy for many types. the only limitation is that only one set of ProxyFuncs is supported per proxy. But this limitation is not a problem, since the proxies can do their own dispatching. For geany it simplifies storing the pointers (avoids introducing a list of vtables, etc)

Okay, so maybe that part is not really relevant indeed.

kugel- commented:
Am 24.09.2015 um 23:42 schrieb Colomban Wendling:
>
>     The separate API allows to do the registration in the plugin's
>     init(), and have the file extensions deppend on plugin
>     configuration. Also a API-wise separattion to
>     geany_plugin_register() is needed to allow for nested proxies.
>
> I get that, my point was that I would find the API less odd if one 
> filled a specific structure, rather than one already used for 
> something else.
>
> e.g.
>
> Proxy proxy = {
>    .extensions  = {"so",NULL  },
>    .load  = my_load,
>    .probe  = my_probe,
>    .unload  = my_unload
> };
> geany_plugin_register_proxy(plugin, &proxy);

Wait a minute, for the new_hooks you convinced me that *this* method 
flawed (due to the points you mentioned) and now you're trying to sell 
it to me? I'm confused.

The way I implemented aimed to be consistent with the just-merged 
new_hooks and other stuff and doesn't have extensibility problems.

Best regards

b4n commented:
> Wait a minute, for the new_hooks you convinced me that *this* method flawed (due to the points you mentioned) and now you're trying to sell it to me? I'm confused.

It's a little more subtle and a little less absurd than that :)

IIRC what you initially did in new_hooks was to fill in a struct to copy it back to the GeanyPlugin structure.  As I agree the funcs are part of the plugin itself, and as it has the above-mentioned flaws, I found it to make a lot more sense to fill the existing structure right away.

Here the thing I find odd is that I feel like the plugin itself and the proxy are two separate things, so filling the first to register the second seem odd to me.  They require calling 2 different register functions, but on the same structure.  That's the only reason why I somewhat try to find an alternative API I find more straightforward (for me).
Do anybody else feel that way?  @elextr @codebrainz ?

Anyway, having
```C
Proxy *proxy = proxy_new();
proxy->probe = my_probe;
...
plugin_registre_proxy(plugin, proxy);
```
would basically make as much sense than the plain struct initializer solution, but requires adding more API (the _new()).  Not saying it's a great API either just now though :)

elextr commented on doc/plugins.dox:
```diff
@@ -725,4 +726,401 @@ void geany_load_module(GeanyPlugin *plugin)
+should respond the same way as for version-incompatible plugins: don't register the plugin at
+all, but leave a message the user suggesting what has to be installed in order to work. Handle
+sytanx errors in the scripts of sub-plugins the same way if possible.
+
```
syntax

elextr commented on doc/plugins.dox:
```diff
@@ -725,4 +726,401 @@ void geany_load_module(GeanyPlugin *plugin)
+The first line acts as a verification that this file is truly a sub-plugin. Within the [Init] section
+there is the menu items for Geany's tools menu. The [Help] section declares the sub-plugins help
+text which is shown in its help dialog (via GeanyPluginFuncs::help). The [Info] section is
+used as-is for filling the sub-plugins PluginInfo fields.
```
I would make this a list with each section a list item rather than one paragraph.

"The [Init] section contains the menu items ..." would be better English.

What is the number after item? Is it the position in the menu?  Is it just a meaningless variation so the keys are different, in which case maybe it would maybe be better as a list rather than separate items, do mnemonic underscores work?

What do you mean "as-is" for the info section?

elextr commented on doc/plugins.dox:
```diff
@@ -725,4 +726,401 @@ void geany_load_module(GeanyPlugin *plugin)
+As already mentioned the file format includes a magic first line which must be present.
+GeanyProxyFuncs::probe() verifies that it's present and avoids showing the sub-plugin in the
+Plugin Manager if not.
+
```
As a general comment, I got confused which functionalities are part of Geany, and which are part of the proxy because the language says a function "does something" making it sound like it exists and so is part of Geany, rather than "must do something" which states a requirement for the proxy.  So the above sentence should read:

"GeanyProxyFuncs::probe() should verify that the correct identification is present and return `false` so Geany avoids showing the sub-plugin in the Plugin Manager."

That way its clear what the callback function does and what Geany does.

This applies in other places as well.

kugel- replied:
In this example section the implementations of the demo proxy are detailed. So the text does in fact describe what the function does (and not what it should do, although it refers to aforementioned requirements).

What the callbacks should do is detailed in an earlier section

elextr commented on doc/plugins.dox:
```diff
@@ -725,4 +726,401 @@ void geany_load_module(GeanyPlugin *plugin)
+problem) which means the sub-plugin cannot be enabled.
+
+It also installs wrapper functions for the inferior's GeanyPluginFuncs as Ini files aren't code.
+It's very likely that your proxy needs something similar because you can only install function
```
s/inferior/sub-plugin/g rather than using mixed terms, sometimes "sub-plugin", sometimes "inferior".

In general in general English "sub-plugin" has fewer negative connotations compared to "inferior plugin" and isn't the main point of these proxy changes to make the "sub-plugins" appear to be normal plugins.

b4n replied:
agreed.  at first "inferior" bugged me, but a dictionary lookup made me shut up and leave it to @elextr, which has now stricken :]

elextr commented on doc/plugins.dox:
```diff
@@ -725,4 +726,402 @@ void geany_load_module(GeanyPlugin *plugin)
+	plugin->proxy_funcs->unload = demoproxy_unload;
+
+	return geany_plugin_register_proxy(plugin, extensions);
+}
```
As discussed on IRC [here](http://irc.geany.org/logs/log_20150924.html), while I understand what @b4n means by the API is "funny" I don't think there is a better one that still meets at least some of simple, extendable, typesafe, so I would suggest to just leave it with the current proposal.

elextr commented on doc/plugins.dox:
```diff
@@ -725,4 +726,402 @@ void geany_load_module(GeanyPlugin *plugin)
+dialog and installs the menu items. proxy_help() is called when the user clicks the help button in
+the Plugin Manager. Consequently, this fires up a suitable dialog, although with a dummy message.
+proxy_cleanup() frees all memory allocated in proxy_init().
+
```
It isn't clear if the ini file is just something this demo uses to define its sub-plugins, or if its required that all proxies use it, and if so why?

kugel- replied:
The entire section is all about the demo proxy. The introductory paragraph tells that the demo proxy loads plugins from ini files for demoing purposes. I don't know how to make it clearer 

Am 25. September 2015 03:43:48 MESZ, schrieb elextr <notifications@github.com>:
>> +	g_free((gchar *)plugin->info->author);
>> +
>> +	g_key_file_free(data->file);
>> +	g_free(data);
>> +}
>> +@endcode
>> +
>> +Finally the demo_proxy's wrapper GeanyPluginFuncs. They are called
>for each possible sub-plugin and
>> +therefore have to multiplex between each using the plugin-defined
>data pointer. Each is called by
>> +Geany as if it were an ordinary, native plugin.
>> +
>> +proxy_init() actually reads the inferior's file using GKeyFile APIs.
>It prepares for the help
>> +dialog and installs the menu items. proxy_help() is called when the
>user clicks the help button in
>> +the Plugin Manager. Consequently, this fires up a suitable dialog,
>although with a dummy message.
>> +proxy_cleanup() frees all memory allocated in proxy_init().
>> +
>
>It isn't clear if the ini file is just something this demo uses to
>define its sub-plugins, or if its required that all proxies use it, and
>if so why?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/629/files#r40393473

kugel- commented:
@b4n @elextr @codebrainz pushed an update, please merge (after Squashing)

b4n commented on plugins/demoproxy.c:
```diff
@@ -0,0 +1,205 @@
+	while (TRUE)
+	{
+		GtkWidget *item;
+		gchar *key = g_strdup_printf(fmt, i++);
```
`fmt` should be the string literal itself.  the variable is not needed, and it prevents GCC from checking the format string.

b4n replied:
(weird, I would think I already mentioned this earlier)

b4n commented on plugins/demoproxy.c:
```diff
@@ -0,0 +1,205 @@
+	PluginContext *data;
+	GtkWidget *dialog;
+	gchar *help_text;
+	GKeyFile *file;
```
`help_text` and `file` are not used

b4n commented on src/plugins.c:
```diff
@@ -747,32 +876,84 @@ plugin_cleanup(Plugin *plugin)
+
+	/* Remove the proxy from the proxy list first. It might appear more than once (once
+	 * for each extension), but if it doesn't appear at all it's not actually a proxy */
+	foreach_list_safe(node, active_proxies.head)
```
prev function does the exact same thing without a macro.  it would be more consistent using the same technique in the 2 places (either one is OK with me, @codebrainz would prefer previous func's one)

kugel- replied:
Which other place?

b4n replied:
https://github.com/geany/geany/pull/629/files#diff-315e8720481183cf75b604b969f170d6R892

b4n commented on src/plugins.c:
```diff
@@ -830,25 +1011,80 @@ static gboolean check_plugin_path(const gchar *fname)
+		return FALSE;
+	/* ensure the dot is really part of the filename */
+	else if (strchr(ext, G_DIR_SEPARATOR) != NULL)
+		return FALSE;
```
Hum, didn't I already asked whether we shouldn't check both `\` and `/` on Windows here too?

kugel- replied:
What can you suggest. Two ifs?

b4n replied:
unfortunately yes, something like adding
```C
#ifdef G_OS_WIN32
	else if (strchr(ext, '/') != NULL)
		return FALSE;
#endif
```

kugel- replied:
BTW, this is always called with the result of g_build_filename() (either directly or via plugin->filename) so it's kind of guaranteed it's a canonical path.

kugel- replied:
Actually, I'm preferring to leave it. We can't have paranoid checks everywhere, even in the deepest leaf functions. At some point we can/have to/should depend on earlier code to provide canonical paths, as is the case here.

b4n replied:
> BTW, this is always called with the result of g_build_filename() (either directly or via plugin->filename) so it's kind of guaranteed it's a canonical path.

Fair enough if it's the case.  Can't this change in the prefs? (e.g. if I manually alter the config file of the enabled plugin list)

> Actually, I'm preferring to leave it. We can't have paranoid checks everywhere, even in the deepest leaf functions. At some point we can/have to/should depend on earlier code to provide canonical paths, as is the case here.

Well, sure, if it's something we can totally rely on.  But that means we need to guarantee it -- and here it's not really a "deep function" in that it extract a very low level info (the extension) --, e.g. that we don't work on paths as the OS understands it, but on an internal representation.

kugel- replied:
In this case it is guaranteed. geany.cfg stores only the plugin's basename, not the full path. The full path is always constructed using g_build_path() prior to calling plugin_new()

b4n replied:
Doesn't seem to be true if I can read it (quickly) correctly: `load_active_plugins()` read `active_plugins_pref` directly, which was loaded from the config file.  There the path is obviously absolute (and it's the case in my local Geany's config), as it's checked for existence.
So AFAICT I could at least force a `\` in the path by manually editing the config file.

b4n commented on src/plugins.c:
```diff
@@ -856,20 +1092,18 @@ static void
+		PluginProxy *proxy = is_plugin(fname);
+
+		if (proxy != NULL && plugin_new(proxy->plugin, fname, FALSE, TRUE))
+				count++;
```
weird indentation

b4n commented on doc/plugins.dox:
```diff
@@ -725,4 +726,403 @@ void geany_load_module(GeanyPlugin *plugin)
+static gboolean proxy_init(GeanyPlugin *plugin, gpointer pdata)
+{
+	PluginContext *data;
+	gchar fmt[] = "item%d";
```
here too for the format thing

b4n commented on src/plugins.c:
```diff
@@ -525,26 +584,78 @@ plugin_load(Plugin *plugin)
+static gpointer plugin_load_gmodule(GeanyPlugin *proxy, GeanyPlugin *subplugin, const gchar *fname, gpointer pdata)
+{
+	GModule *module;
+	GeanyData **p_geany_data;
```
unused

b4n commented on src/plugins.c:
```diff
@@ -1356,7 +1748,7 @@ static void on_pm_tree_filter_entry_icon_release_cb(GtkEntry *entry, GtkEntryIco
 
 
-static void pm_prepare_treeview(GtkWidget *tree, GtkListStore *store)
+static void pm_prepare_treeview(GtkWidget *tree, GtkTreeStore *store)
```
changes to this function lead to `iter` and `list` variables to be unused.

b4n commented:
Note to myself: don't forget to bump API when merging this!

kugel- commented:
BTW if anyone wants to actually test this, there is the demoproxy (obviously) but also an adapted geanypy on my repo

https://github.com/kugel-/geany-plugins/tree/geanypy-pluxy

b4n commented:
(I quickly tried to start squashing commits, and it'll be a little hard because many fixes alter next commit/fixes a bit too much for automatic merges to work.  Doing it will take some time)

@kugel- thanks for the link, it'll be interesting to see a real usage :)

kugel- commented:
You're welcome. I posted it already in the very early days of the PR :-) 

The Squashing turns our non-trivial, I'll take care

kugel- commented:
Here is the merged branch: https://github.com/kugel-/geany/tree/pluxy-merge

I added commits, which should be self-explainatory. Other than that, I fixed a @since tag and a last "pluxies" occurence.

Tell me if I should push that -merge branch to this PR.

kugel- commented:
I also added the above mengtioned as commits here

b4n commented on src/plugins.c:
```diff
@@ -712,6 +805,40 @@ static void remove_sources(Plugin *plugin)
 
 
+/* returns true if the plugin is native, i.e. not loaded through a proxy.
+ * This determines whether it's backed by a GModule */
```
this comment looks wrong

b4n commented on src/plugins.c:
```diff
@@ -830,25 +1012,87 @@ static gboolean check_plugin_path(const gchar *fname)
+	foreach_list(node, active_proxies.head)
+	{
+		PluginProxy *proxy = node->data;
+		if (utils_str_casecmp(ext, proxy->extension) != 0)
```
This test is inverted, should be `==`  It only "works" with C plugins by chance because they ship with a `.la` and for some reason loading this one also works.
Doesn't work with [geanypy port](https://github.com/kugel-/geany-plugins/commit/9c27df004f1c0906970da605d6c0ffd89d0a2fed), no python plugins show up -- and geanypy probes and tried to load all C plugins.

kugel- commented:
@b4n pushed fixes. I updated https://github.com/kugel-/geany/tree/pluxy-merge accordingly.

b4n commented:
LGTM, and merge branch is in sync.  @codebrainz could you give it a quick test please?

b4n commented:
Pre-rebase commits were:

* 0def4f5 `plugins: introduce load and unload functions for plugins`
* d11ada3 `plugins: generic load_data instead of module pointer in Plugin struct`
* bf7bbca `plugins: refactor GtkListStore population code into separate function`
* 7acda46 `plugins: when loading active ones, loop until no more proxy plugins are added`
* 533c4d6 `plugins: introduce probe() for proxy plugins`
* a006f71 `plugins: add geany_plugin_register_proxy() to the plugin API`
* ce9fe0b `plugins: reselect when toggling the current plugin`
* 0e9923e `demopluxy: add a demo pluxy showcasing how to create a proxy plugin`
* e1e8349 `plugins: improve PM dialog for proxy and sub-plugins`
* 883ac8b `Fixup 0def4f5 (plugins: introduce load and unload functions for plugins)`
* 2e36527 `Fixup d11ada3 (plugins: generic load_data instead of module pointer in Plugin struct)`
* dac823b `Fixup bf7bbca (plugins: refactor GtkListStore population code into separate function)`
* 290e3ab `Fixup 7acda46 (plugins: when loading active ones, loop until no more proxy plugins are added)`
* 50aecf7 `Fixup a006f71 (plugins: add geany_plugin_register_proxy() to the plugin API)`
* 04e5e52 `Fixup ce9fe0b (plugins: reselect when toggling the current plugin)`
* 9f67844 `Fixup 0e9923e (demopluxy: add a demo pluxy showcasing how to create a proxy plugin)`
* 31d94e7 `Fixup e1e8349 (plugins: improve PM dialog for proxy and sub-plugins)`
* d148920 `plugins: use GQueue to restore GLib compatibility`
* 85a1674 `Fixup 0def4f5 #2 (plugins: introduce load and unload functions for plugins)`
* 2e443d1 `Fixup 7acda46 #2 (plugins: when loading active ones, loop until no more proxy plugins are added)`
* b87c711 `Fixup a006f71 #2 (plugins: add geany_plugin_register_proxy() to the plugin API)`
* f86877f `Fixup ce9fe0b #2 (plugins: reselect when toggling the current plugin)`
* 336f641 `Fixup 0e9923e #2 (demopluxy: add a demo pluxy showcasing how to create a proxy plugin)`
* 367f303 `Fixup e1e8349 #2 (plugins: improve PM dialog for proxy and sub-plugins)`
* 5a3872a `Fixup d148920 (plugins: use GQueue to restore GLib compatibility)`
* 6e1dfd4 `Fixup a006f71 #3 (plugins: add geany_plugin_register_proxy() to the plugin API)`
* 8e7de03 `Fixup 0e9923e #3 (demopluxy: add a demo pluxy showcasing how to create a proxy plug`
* f90bc9c `Fixup a006f71 #4 (plugins: add geany_plugin_register_proxy() to the plugin API)`
* 7c329d0 `Fixup 0e9923e #4 (demopluxy: add a demo pluxy showcasing how to create a proxy plugin)`
* 1529158 `Fixup d148920 #2 (plugins: use GQueue to restore GLib compatibility)`
* c54735e `Fixup 0def4f5 #3 (plugins: introduce load and unload functions for plugins)`
* 8a9c66a `Fixup bf7bbca #2 (plugins: refactor GtkListStore population code into separate function)`
* d6978be `Fixup 0e9923e #5 (demopluxy: add a demo pluxy showcasing how to create a proxy plugin)`
* 38b4c03 `Fixup 0def4f5 #4 (plugins: introduce load and unload functions for plugins)`
* be02a70 `Fixup d11ada3 #2 (plugins: generic load_data instead of module pointer in Plugin struct)`
* c7ef4aa `Fixup bf7bbca #3 (plugins: refactor GtkListStore population code into separate function)`
* 13699be `Fixup a006f71 #5 (plugins: add geany_plugin_register_proxy() to the plugin API)`
* df6569a `Fixup a006f71 #6 (plugins: add geany_plugin_register_proxy() to the plugin API)`
* 0252408 `plugins: do not pass potentially destroyed data to unload callback`
* d5a06a7 `plugins: enfore geany_plugin_register_proxy() can be called once`
* 8161913 `Fixup 0def4f5 #5 (plugins: introduce load and unload functions for plugins)`
* efc4f78 `Fixup d11ada3 #3 (plugins: generic load_data instead of module pointer in Plugin struct)`
* 90c905a `Revert "plugins: do not pass potentially destroyed data to unload callback"`
